source/uwp/Renderer/lib/ActionHelpers.cpp

```diff
@@ -753,677 +545,533 @@ namespace AdaptiveCards::Rendering::Uwp::ActionHelpers
 
             // Register the overflow button with the render context
-            ComPtr<IUIElement> overflowButtonAsUIElement;
-            overflowButton.As(&overflowButtonAsUIElement);
 
-            ComPtr<AdaptiveCards::Rendering::Uwp::AdaptiveRenderContext> contextImpl =
-                PeekInnards<AdaptiveCards::Rendering::Uwp::AdaptiveRenderContext>(renderContext);
+            auto contextImpl = peek_innards<rtrender::implementation::AdaptiveRenderContext>(renderContext);
 
             if (adaptiveActionSet != nullptr)
             {
-                RETURN_IF_FAILED(contextImpl->AddOverflowButton(adaptiveActionSet, overflowButtonAsUIElement.Get()));
+                // TODO: no need to .as<rtxaml::UIElement> for overflowButton, correct?
+                contextImpl->AddOverflowButton(adaptiveActionSet, overflowButton);
             }
             else
             {
-                RETURN_IF_FAILED(contextImpl->AddOverflowButton(adaptiveCard, overflowButtonAsUIElement.Get()));
+                contextImpl->AddOverflowButton(adaptiveCard, overflowButton);
             }
         }
 
         // Reset icon placement value
-        RETURN_IF_FAILED(renderArgs->put_AllowAboveTitleIconPlacement(false));
+        renderArgs.AllowAboveTitleIconPlacement(false);
+        XamlHelpers::SetStyleFromResourceDictionary(renderContext, {L"Adapative.Actions"}, actionsPanel);
 
-        ComPtr<IFrameworkElement> actionsPanelAsFrameworkElement;
-        RETURN_IF_FAILED(actionsPanel.As(&actionsPanelAsFrameworkElement));
-        RETURN_IF_FAILED(XamlHelpers::SetStyleFromResourceDictionary(renderContext,
-                                                                     L"Adaptive.Actions",
-                                                                     actionsPanelAsFrameworkElement.Get()));
-
-        ComPtr<IStackPanel> actionSet =
-            XamlHelpers::CreateABIClass<IStackPanel>(HStringReference(RuntimeClass_Windows_UI_Xaml_Controls_StackPanel));
-        ComPtr<IPanel> actionSetAsPanel;
-        actionSet.As(&actionSetAsPanel);
+        rtxaml::Controls::StackPanel actionPanel;
 
         // Add buttons and show cards to panel
-        XamlHelpers::AppendXamlElementToPanel(actionsPanel.Get(), actionSetAsPanel.Get());
-        XamlHelpers::AppendXamlElementToPanel(showCardsStackPanel.Get(), actionSetAsPanel.Get());
+        XamlHelpers::AppendXamlElementToPanel(actionsPanel, actionPanel);
+        XamlHelpers::AppendXamlElementToPanel(showCardsStackPanel, actionPanel);
 
-        return actionSetAsPanel.CopyTo(actionSetControl);
+        return actionPanel;
     }
 
-    void CreateAppropriateButton(ABI::AdaptiveCards::ObjectModel::Uwp::IAdaptiveActionElement* action, ComPtr<IButton>& button)
+    rtxaml::Controls::Button CreateAppropriateButton(rtom::IAdaptiveActionElement const& action)
     {
-        if (action != nullptr)
+        if (action && (action.ActionType() == rtom::ActionType::OpenUrl))
```

Comment:
```suggestion
        if (action.ActionType() == rtom::ActionType::OpenUrl)
```

alternatively, you could use a ternary